### ISSUE TITLE
Firewalld fails to load IP sets with overlapping entries

### ISSUE THREAD

**Issue by vgaetera** [firewalld/firewalld#1343]
### What happened

Firewalld fails to load the entire IP set if it contains overlapping entries.
The problem escalates when using IP intervals for entries.
This is a major barrier to efficiently using IP sets in Firewalld.

### What you expected to happen

Overlapping entries is small issue that deserves a warning at most.
It should not be escalated to a critical error that affects the entire setup.

IP sets are commonly used for policy-based routing and traffic filtering.
It can be generated from GeoIP databases and contain thousands of entries.

Firewalld should tolerate overlapping IP set entries.
Overlapping entries should not prevent the IP set from loading.

##### Example 1
Firewalld fails to populate an IP set from a file:
```sh
tee << EOF > /tmp/test.ipset
192.168.0.0/24
192.168.0.1/32
EOF
sudo firewall-cmd --permanent --delete-ipset=test -q
sudo firewall-cmd --permanent --new-ipset=test --type=hash:net --family=inet
sudo firewall-cmd --permanent --ipset=test --add-entries-from-file=/tmp/test.ipset
sudo firewall-cmd --permanent --info-ipset=test
sudo cat /etc/firewalld/ipsets/test.xml
```

##### Example 2
Firewalld fails to reload and returns a JSON blob:
```sh
sudo firewall-cmd --permanent --delete-ipset=test -q
sudo firewall-cmd --permanent --new-ipset=test --type=hash:net --family=inet
sudo firewall-cmd --permanent --ipset=test --add-entry=192.168.0.0-192.168.0.1
sudo firewall-cmd --permanent --ipset=test --add-entry=192.168.0.1-192.168.0.2
sudo firewall-cmd --reload
```

### Firewalld Version

2.1.2

### Firewalld Backend

nftables

### Linux distribution

Fedora Linux 40 (Forty)

### Linux kernel version

6.8.9-300.fc40.x86_64

**Comment by erig0** [comment]
> Overlapping entries is a minor issue that deserves a warning at most.

It's not that simple. firewalld supports multiple backends each with their own nuances and bugs around coalescing and breaking CIDR/range entries. Adding overlapping sets is usually okay. It because complicated when you start removing entries.

e.g.
add 10.10.10.0/24
add 10.10.0.0/16
add 10.0.0.0/8

then
remove 10.10.0.0/16

On the backend "10.10.10.0/24" would also be implicitly removed. That means firewalld has to also detect and handle this.

Or maybe the user meant to "remove" everything in 10.10.0.0/16, EXCEPT those in 10.10.10.0/24. This is harder, firewalld would have to figure it out and only send the real removes to the backends. That gets complicated for multi dimensional sets.

In short, it's much less complicated to disallow overlapping sets.

---

It may be that the backends have better support for breaking and coalescing entries and this can be revisited. It's important to note that disallowing overlapping entries was done in response to bugs and reports.

**Comment by vgaetera** [comment]
Thanks for the answer, I hope this can be implemented someday, otherwise it requires to find and eliminate possible overlapping entries from the list before trying to load it with Firewalld, which is not so trivial specifically when working with large IP sets utilizing GeoIP data.

***
Should we split the Example#2 to a separate ticket?
It results in the following ruleset:
```sh
> nft list ruleset
table inet firewalld_policy_drop {
	chain filter_input {
		type filter hook input priority filter + 9; policy drop;
		ct state { established, related } accept
		drop
	}

	chain filter_forward {
		type filter hook forward priority filter + 9; policy drop;
		ct state { established, related } accept
		drop
	}

	chain filter_output {
		type filter hook output priority filter + 9; policy drop;
		ct state { established, related } accept
		drop
	}
}
table inet firewalld {
}
```

**Comment by msaladna** [comment]
We can put a man on the moon and yet modern government contracts kill this project?

**Comment by Hedrauta** [comment]
TLDR for someone similiar issues with this kind of error:

Check your ipset, if you have "double entries" ( X.X.X./22 and X.X.X./24 for example ) and remove the one with lesser priority

|

i had a similiar issue with my hash:net ipset-file...

took me 3 hrs and a psychological meltdown until i got it back running... god damn am i happy, that my SSH-Connection wasnt killed.
